drivers/node/vsphere/vsphere.go

```diff
@@ -0,0 +1,248 @@
+package vsphere
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/portworx/sched-ops/task"
+	"github.com/portworx/torpedo/drivers/node"
+	"github.com/portworx/torpedo/drivers/node/ssh"
+	"github.com/sirupsen/logrus"
+	"github.com/vmware/govmomi"
+	"github.com/vmware/govmomi/find"
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25/types"
+)
+
+const (
+	// DriverName is the name of the vsphere driver
+	DriverName = "vsphere"
+	// Protocol is the protocol used
+	Protocol = "https://"
+)
+
+const (
+	vsphereUname = "VSPHERE_USER"
+	vspherePwd   = "VSPHERE_PWD"
+	vsphereIP    = "VSPHERE_HOST_IP"
+)
+
+const (
+	// DefaultUsername is the default username used for ssh operations
+	DefaultUsername = "root"
+	// VMReadyTimeout Timeout for checking VM power state
+	VMReadyTimeout = 3 * time.Minute
+	// VMRetryInterval interval for retry when checking VM power state
+	VMRetryInterval = 5 * time.Second
+)
+
+// Vsphere ssh driver
+type vsphere struct {
+	ssh.SSH
+	vsphereUsername string
+	vspherePassword string
+	vsphereHostIP   string
+	ctx             context.Context
+	cancel          context.CancelFunc
+}
+
+var (
+	vmMap = make(map[string]*object.VirtualMachine)
+)
+
+func (v *vsphere) String() string {
+	return DriverName
+}
+
+// InitVsphere initializes the vsphere driver for ssh
+func (v *vsphere) Init() error {
+	logrus.Infof("Using the vsphere node driver")
+	v.SSH.Init()
+
+	username := os.Getenv(vsphereUname)
+	if len(username) == 0 {
+		v.vsphereUsername = DefaultUsername
+	} else {
+		v.vsphereUsername = username
+	}
+
+	v.vspherePassword = os.Getenv(vspherePwd)
+	if len(v.vspherePassword) == 0 {
+		return fmt.Errorf("Vsphere password not provided as env var: %s", vspherePwd)
+	}
+
+	v.vsphereHostIP = os.Getenv(vsphereIP)
+	if len(v.vsphereHostIP) == 0 {
+		return fmt.Errorf("Vsphere host IP not provided as env var: %s", vsphereIP)
+	}
+	err := v.connect()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// TestConnection tests the connection to the given node
+func (v *vsphere) TestConnection(n node.Node, options node.ConnectionOpts) error {
+	var err error
+	logrus.Infof("Testing vsphere driver connection by checking state of the VMs in the vsphere")
+	if _, ok := vmMap[n.Name]; !ok {
+		return fmt.Errorf("Failed to get VM: %s", n.Name)
+	}
+	vm := vmMap[n.Name]
+
+	t := func() (interface{}, bool, error) {
+		powerState, err := vm.PowerState(v.ctx)
+		if err != nil || powerState != types.VirtualMachinePowerStatePoweredOn {
+			return nil, true, &node.ErrFailedToRebootNode{
+				Node:  n,
+				Cause: fmt.Sprintf("Failed to test connection to VM: %s Current Status: %v, error: %v", vm.Name(), powerState, err),
+			}
+		}
+		return nil, false, nil
+	}
+	if _, err := task.DoRetryWithTimeout(t, VMReadyTimeout, VMRetryInterval); err != nil {
+		return err
+	}
+	return err
+}
```

Comment:
The reason for this method is not just check if node is up, but also it is running. The VM could be in boot process and can't execute any commands. Add the execution of simple command like date or hostname as we use in ssh driver.

Reply:
Added execution of hostname command once it's verified that VM is powered on.